spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/ManagedCommitSuite.scala

```diff
@@ -116,6 +123,405 @@ class ManagedCommitSuite
+        DeltaLog.clearCache()
+
+        // Add new commit to convert FS table to managed-commit table
+        val deltaLog2 = DeltaLog.forTable(spark, tablePath)
+        val oldMetadata = snapshotV2.metadata
+        val newMetadata = snapshotV2.metadata.copy(
+          configuration =
+            oldMetadata.configuration + (MANAGED_COMMIT_OWNER_NAME.key -> "tracking-in-memory"))
+        deltaLog2.startTransaction().commitManually(newMetadata)
+        commitStore.registerTable(deltaLog2.logPath, 3)
+        deltaLog2.startTransaction().commitManually(createTestAddFile("f2"))
+        deltaLog2.startTransaction().commitManually()
+        val snapshotV5 = deltaLog2.unsafeVolatileSnapshot
+        assert(snapshotV5.version === 5)
+        assert(snapshotV5.commitStoreOpt.nonEmpty)
+        // only delta 4/5 will be un-backfilled and should have two dots in filename (x.uuid.json)
+        assert(snapshotV5.logSegment.deltas.count(_.getPath.getName.count(_ == '.') == 2) === 2)
+
+        val usageRecords = Log4jUsageLogger.track {
+          val newSnapshotV5 = deltaLog1.update()
+          assert(newSnapshotV5.version === 5)
+          assert(newSnapshotV5.logSegment.deltas === snapshotV5.logSegment.deltas)
+        }
+        assert(filterUsageRecords(usageRecords, "delta.readChecksum").size === 2)
+      }
+    } finally {
+      spark.conf.set(MANAGED_COMMIT_OWNER_NAME.defaultTablePropertyKey, oldCommitOwnerValue)
+    }
+  }
+
+  test("update works correctly with InitialSnapshot") {
+    CommitStoreProvider.registerBuilder(TrackingInMemoryCommitStoreBuilder(batchSize = 2))
+    withTempDir { tempDir =>
+      val tablePath = tempDir.getAbsolutePath
+      val clock = new ManualClock(System.currentTimeMillis())
+      val log = DeltaLog.forTable(spark, new Path(tablePath), clock)
+      assert(log.unsafeVolatileSnapshot.isInstanceOf[InitialSnapshot])
+      assert(log.getCapturedSnapshot().updateTimestamp == clock.getTimeMillis())
+      clock.advance(500)
+      log.update()
+      assert(log.unsafeVolatileSnapshot.isInstanceOf[InitialSnapshot])
+      assert(log.getCapturedSnapshot().updateTimestamp == clock.getTimeMillis())
+    }
+  }
+
+  // This test has the following setup:
+  // Setup:
+  // There are two table owners: CS1 and CS2 both of which are pointing to same underlying
+  // in-memory implementation.
+  // 1. Make 3 commits on the table with CS1 as owner.
+  // 2. Modify the content of commit-2. Change the table owner from 1 to 2 as part of it.
+  // 3. Do cold read from table and confirm we can construct snapshot v2 automatically. This will
+  //    need recursive snapshot update and both CS1 and CS2 will be contacted one after the other.
+  // 4. Write commit 3/4 using new commit owner.
+  // 5. Read the table again and make sure right APIs are called:
+  //    a) If read query is run in scala, we do listing 2 times. So CS2.getCommits will be called
+  //       twice. We should not be contacting CS1 anymore.
```

Comment:
I guess we expect backfill somewhere along the way? But when does that occur, and wouldn't commit-2 and its backfilled counterpart disagree after we modify?

Reply:
Update: Actually, batch size 10 means we don't backfill anything, and instead rely on CS2 to know about commits performed by CS1, and also rely on warm start snapshot update to start directly from CS2?

Reply:
> Actually, batch size 10 means we don't backfill anything, and instead rely on CS2 to know about commits performed by CS1, and also rely on warm start snapshot update to start directly from CS2

Yes - this is correct. Even if we add a constrain in RFC that commits must be backfilled on ownership change, still our delta-spark code could recursively update snapshots across different owners even when the commits where ownership changes are not backfilled. This test case is testing the same.

Reply:
> Even if we add a constrain in RFC that commits must be backfilled on ownership change, still our delta-spark code could recursively update snapshots across different owners even when the commits where ownership changes are not backfilled. This test case is testing the same.

Why would we need to validate that scenario, once the spec forbids it?

Reply:
That scenario has been removed.